x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexSearcherWrapper.java

```diff
@@ -120,47 +91,23 @@ protected DirectoryReader wrap(DirectoryReader reader) {
+            final IndicesAccessControl.IndexAccessControl permissions = indicesAccessControl.getIndexPermissions(shardId.getIndexName());
             // No permissions have been defined for an index, so don't intercept the index reader for access control
             if (permissions == null) {
                 return reader;
             }
 
-            if (permissions.getQueries() != null) {
-                BooleanQuery.Builder filter = new BooleanQuery.Builder();
-                for (BytesReference bytesReference : permissions.getQueries()) {
-                    QueryShardContext queryShardContext = queryShardContextProvider.apply(shardId);
-                    String templateResult = evaluateTemplate(bytesReference.utf8ToString());
-                    try (XContentParser parser = XContentFactory.xContent(templateResult)
-                            .createParser(queryShardContext.getXContentRegistry(), LoggingDeprecationHandler.INSTANCE, templateResult)) {
-                        QueryBuilder queryBuilder = queryShardContext.parseInnerQueryBuilder(parser);
-                        verifyRoleQuery(queryBuilder);
-                        failIfQueryUsesClient(queryBuilder, queryShardContext);
-                        Query roleQuery = queryShardContext.toQuery(queryBuilder).query();
-                        filter.add(roleQuery, SHOULD);
-                        if (queryShardContext.getMapperService().hasNested()) {
-                            NestedHelper nestedHelper = new NestedHelper(queryShardContext.getMapperService());
-                            if (nestedHelper.mightMatchNestedDocs(roleQuery)) {
-                                roleQuery = new BooleanQuery.Builder()
-                                    .add(roleQuery, FILTER)
-                                    .add(Queries.newNonNestedFilter(queryShardContext.indexVersionCreated()), FILTER)
-                                    .build();
-                            }
-                            // If access is allowed on root doc then also access is allowed on all nested docs of that root document:
-                            BitSetProducer rootDocs = queryShardContext.bitsetFilter(
-                                    Queries.newNonNestedFilter(queryShardContext.indexVersionCreated()));
-                            ToChildBlockJoinQuery includeNestedDocs = new ToChildBlockJoinQuery(roleQuery, rootDocs);
-                            filter.add(includeNestedDocs, SHOULD);
-                        }
-                    }
+            DirectoryReader wrappedReader = reader;
+            DocumentPermissions documentPermissions = permissions.getDocumentPermissions();
+            if (documentPermissions != null && documentPermissions.hasDocumentLevelPermissions()) {
+                BooleanQuery filterQuery = DocumentPermissions.filter(getUser(), scriptService, shardId, queryShardContextProvider,
+                        documentPermissions);
```

Comment:
It's a little bit of a personal preference, but it feels to me like this should be an instance method that's called on `documentPermissions` rather than a static method.

Reply:
It still looks like a `static` method to me.
I think we want:
```
BooleanQuery filterQuery = documentPermissions.filter(getUser(), scriptService, shardId, queryShardContextProvider);
```